render-ws-spark-client/src/main/java/org/janelia/render/client/spark/newsolver/DistributedAffineBlockSolverClient.java

```diff
@@ -344,48 +349,64 @@ private static void globallySolveOneSetup(final List<AffineBlockSolverSetup> set
-
-        while (localIterator.hasNext()) {
-
-            final Tuple2<Integer, BlockData<AffineModel2D, ?>> tuple2 = localIterator.next();
-            final int setupIndex = tuple2._1;
-            final BlockData<AffineModel2D, ?> outputBlock = tuple2._2;
-
-            if ((previousSetupIndex != null) && (previousSetupIndex != setupIndex)) {
-                globallySolveOneSetup(setupList, previousSetupIndex, solverList, outputBlocksForSetup);
-                outputBlocksForSetup.clear();
-            }
-
-            outputBlocksForSetup.add(outputBlock);
-            previousSetupIndex = setupIndex;
-        }
-
-        if (outputBlocksForSetup.isEmpty()) {
-            throw new IOException("no blocks were computed for setupIndex " + previousSetupIndex +
-                                  ", something is wrong");
-        } else {
-            globallySolveOneSetup(setupList, previousSetupIndex, solverList, outputBlocksForSetup);
-        }
+        // 2. Combine the solved output blocks for each setup (stack) into a list so that the list of blocks
+        //    can be solved globally on a Spark executor.  Doing the global solve on executor/workers allows
+        //    global solves for different stacks to be run concurrently.
+        final JavaPairRDD<Integer, List<BlockData<AffineModel2D, ?>>> outputBlocksForSetupRdd =
+                rddOutputBlocks.combineByKey(
+                        // createCombiner
+                        block -> {
+                            final List<BlockData<AffineModel2D, ?>> list = new ArrayList<>();
+                            list.add(block);
+                            return list;
+                        },
+                        // mergeValue
+                        (list, block) -> {
+                            list.add(block);
+                            return list;
+                        },
+                        // mergeCombiners
+                        (list1, list2) -> {
+                            list1.addAll(list2);
+                            return list1;
+                        }
+                );
+
+        // 3. Run the global solve for each setup (stack) in parallel.
+        final JavaRDD<StackId> globallySolvedTargetStackIdsRdd =
+                outputBlocksForSetupRdd.map(setupIndexWithOutputBlocks -> {
+
+                    final int setupIndex = setupIndexWithOutputBlocks._1;
+                    final List<BlockData<AffineModel2D, ?>> outputBlocks = setupIndexWithOutputBlocks._2;
+
+                    LogUtilities.setupExecutorLog4j("setupIndex" + setupIndex); // add setup index to log context
+
+                    globallySolveOneSetup(setupList, setupIndex, solverList, outputBlocks);
+
+                    final AffineBlockSolverSetup setup = setupList.get(setupIndex);
+                    return new StackId(setup.targetStack.owner, setup.targetStack.project, setup.targetStack.stack);
+                });
```

Comment:
This is a pretty long lambda. Could this be a private method?

Reply:
sure